docs/tendermint-core/consensus/proposer-based-timestamps.md

```diff
@@ -0,0 +1,89 @@
+---
+order: 3
+---
+
+# Proposer-Based Timestamps
+
+ This document provides an overview of the Proposer-Based Timestamp algorithm 
+ added to Tendermint in the v0.36 release. This document outlines the core functionality
+ as well as the parameters and constraints of the this algorithm.
+
+# Algorithm Overview 
+
+The proposer-based timestamps algorithm defines a way for a Tendermint blockchain
+to create block timestamps that are within a reasonable bound of the clocks of the
+validators on the network.
+
+## Algorithm Parameters
+
+The functionality of the Proposer-Based Timestamps algorithm is governed by two
+parameters within Tendermint. These two parameters are [consensus parameters](https://github.com/tendermint/tendermint/blob/c42c6d06d2986791ad1cb8742824818edd9b6323/spec/abci/apps.md#L291), 
+meaning they are configured by the ABCI application and are expected to be the
+same across all nodes on the network.
+
+### `Precision`
+
+The `Precision` parameter configures the acceptable upper-bound of clock drift among
+all of the nodes on a Tendermint network. Any two nodes on a Tendermint network
+are expected to have clocks that differ by at most `Precision` seconds? Ms? at
+any given instant.
+
+### `MessageDelay`
+
+The `MessageDelay` parameter configures the acceptable upper-bound for transmitting
+a `Proposal` message from the proposer to _all_ of the validators on the network.
```

Comment:
Do we really mean "all" here, or "each"? I.e., is it the worst case among all my peers, or the sum of all the times?

Reply:
Worst case among all peers. I think that would be all?

Reply:
Worst case among all peers is sufficient presuming you can message all peers in parallel. That seems like an implementation dependent property: if the process can't open enough ports or the link is saturated, the time may exceed the longest individual latency.

So I guess this means "all", which means in the purely hypothetical case where only one peer can be contacted at a time,  would be the sum (in the limit).

Reply:
> if the process can't open enough ports or the link is saturated, the time may exceed the longest individual latency.

Well, there's the additional wrinkle that network topologies are not complete graphs. So, the hope here is that we communicate the understanding that the message needs to reach all of the other nodes on the network.